origami-parsers.el

```diff
@@ -767,6 +767,88 @@ See function `origami-python-parser' description for argument CREATE."
+                                 ;; attribute value
+                                 (seq "\""
+                                      (zero-or-more (not "\""))
+                                      "\"")))
+                               ">"))
+                         (end-tag
+                          (seq "</" (any word "_") (zero-or-more (not ">")) ">")))
+                      ;; no need to care for comments/CDATA, these pos are filtered by face
+                      ;; in valid-pos-p
+                      (let* ((rx-beg-tag (rx beg-tag))
+                             (rx-end-tag (rx end-tag))
+                             (rx-all (rx (or beg-tag end-tag)))
+                             (positions (origami-get-positions content rx-all #'valid-pos-p)))
+                        (origami-build-pair-tree create rx-beg-tag rx-end-tag nil positions))))
+       (trim (nodes)
+             ;; trims leaves, if required
+             ;; no artificial root node yet, so base level is a list of nodes
+             (if (not remove-leaves)
+                 nodes
+               (let ((trimmed-nodes
+                      ;; remove base level leaves
+                      (seq-remove (lambda (node) (null (origami-fold-children node)))
+                                  nodes)))
+                 (if (seq-empty-p trimmed-nodes)
+                     trimmed-nodes
+                   ;; recurse to remove deeper level leaves
+                   (-map #'trim-recurse trimmed-nodes)))))
+       (trim-recurse (tree)
+                     (if (origami-fold-children tree)
+                         ;; has children - recurse for children, and don't remove this node
+                         (let ((trimmed-childs
+                                (-map #'trim-recurse (origami-fold-children tree))))
+                           ;; return copy of node with filtered childs
+                           (origami-fold-children-set tree
+                                                      (seq-remove #'null trimmed-childs)))
+                       ;; no childrens, remove this node
+                       nil)))
+    (lambda (content)
+      (-some-> content
+        build-nodes
+        trim))))
+
+(defcustom origami-xml-skip-leaf-nodes t
+  "In xml files, only elements with child elements will be
+foldable, not if they contain text only."
+  :type 'boolean
+  :group 'origami)
+
+(defun origami-xml-parser (create)
+  "Parser for xml."
+  (origami-xml-base-parser create origami-xml-skip-leaf-nodes))
+
+(defun origami-html-parser (create)
+  "Parser for html."
+  (rx-let ((ignore-tags (&rest tags) (seq "<" (or tags) word-end)))
+    ;; Self-closing tags (void elements) without closing slash would throw off parser, ignore
+    (let ((ignore-tags-rx
+           (rx (ignore-tags "area" "base" "br" "col" "command" "embed" "hr" "img" "input"
+                            "keygen" "link" "menuitem" "meta" "param" "source" "track" "wbr"))))
```

Comment:
I might extract this to a custom variable. WDYT?

Reply:
Two points I had in mind here: Avoid regex construction on each parser call, and html void elements are part of a specification, so will not change. But first point won't make a significant difference I guess (my grasp on what happens when in elisp is still lacking), and for the second, someone might parse other formats that are intermixed with html, where customization might be useful. So sure, why not!

Reply:
> Avoid regex construction on each parser call

I think the parser call are only occurs when a buffer triggers origami. I don't think this would be the problem. :)

> html void elements are part of a specification, so will not change

Could be true. I think having an extracted list would easier for me to manage.